Remove bech32 PubKey support

robert-zaremba opened the pull request:
## Description
- [x] removing bech32 pubkeys functions
- [x] remove bech32 from keyring
- [x] removing code duplication

Closes: #7447

This task essentially removes bech32 for public keys in the storage and user layer (clients, outputs etc..).
We are NOT breaking the legacy REST API because it deprecated anyway and will be removed.

### Summary

+ removed bech32 from user facing code
+ updated the keyring display structure
+ changed user/client pubkey serialization
+ added new tests for `ed25519` as an example on how to properly serialize interfaces (both binary and JSON).
+ few API breaking changes: 
+ Created `NewJSONAnyMarshaler` - a helper method to use `codec.(Un)marsjalJSON` withing a client context:
```
am := codec.NewJSONAnyMarshaler(ctx.JSONMarshaler, ctx.InterfaceRegistry)
err := codec.UnmarshalIfcJSON(am, &pk, []byte(pubKey))
```


---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

aaronc commented on x/staking/teststaking/handler.go:
```diff
@@ -0,0 +1,64 @@
+
+// HandlerT is a structure which wraps the staking handler
+// and provides methods useful in tests
+type HandlerT struct {
```
I see what you're getting at here. Maybe there are cleaner ways to do this but this is one approach. I think the naming is confusing. How about just calling this something like `TestHelper`?

robert-zaremba replied:
`TestHelper` is too generic. I'm looking for a name which would represent something like a _handler service_. What I'm doing here is in fact a service for the module.

alessio replied:
The naming is confusing, though `TestHelper` does not help much (and among the two, `HandlerT` sounds better to me). I am definitely more concerned about the package's name - `teststaking` sounds imperative and makes very little sense. I'd suggest to rename it to just `testing`, e.g. the following looks much better from client developer's perspective IMHO: 

```golang
import stakingTesting "github.com/cosmos/cosmos-sdk/x/staking/testing"
```

robert-zaremba replied:
Re package rename. `testing` is a name of the standard (builtin) package, and shouldn't be used otherwise. 
This package is meant to use in tests, to provide helpers / mocks / stubs. I already went through it in past projects and it's good to group them under the package we will mock or provide those helpers.

Any other, shorter name will cause that we will have name conflicts and we will do same things as in other places - package aliases. This doesn't help because we will use it probably everywhere (to bring clarity) and our editors don't like it (because we need to add this packages + aliases manually).

robert-zaremba replied:
Re struct name, I renamed it to `Service`, because as I wrote in the first comment, it's meant to support module services. What do you think?

amaurymartiny replied:
How about `github.com/cosmos/cosmos-sdk/x/staking/testutil`? We already have some of those in the SDK.

aaronc replied:
> Re package rename. `testing` is a name of the standard (builtin) package, and shouldn't be used otherwise.

Agreed

> Any other, shorter name will cause that we will have name conflicts and we will do same things as in other places - package aliases.

While this is true we don't do it in general, for instance we use `x/bank/types` not `x/bank/banktypes`. Going in that direction would be a larger refactor for the SDK that I'm not sure would support readability. (I do btw agree with the reasoning for `legacytx` but that's sort of an isolated case IMHO.)

I support `testutil.`

aaronc replied:
> Re struct name, I renamed it to `Service`, because as I wrote in the first comment, it's meant to support module services. What do you think?

How about `HandlerHelper` or `ServiceHelper`? I just think `Helper` makes it clearer it's for tests.

robert-zaremba replied:
`test` is already in package name, so I don't think we need to use helper suffixes (neither in package name nor strict name). 
`testutil.Service` or `teststaking.Service` is short and clear (Go spirit)
Personally I prefer the latter.
The argument against the `testutil` is same as in my comments above: unnecessary generic. This package doesn't provide general testing utilities. If we will adopt that name then we will only increase the package aliases usage, which IMHO is not good (I explained it in my earlier comments)

robert-zaremba replied:
However if it's only me who doesn't like the `testutil` name, then ok, I will rename (ps, @alexanderbez  we were agreeing that package names like util, utils, common ... are not good, testutil is not better IMHO).

marbar3778 commented:
It may be good to keep the removal of bech32 and removing code duplication/test helper package as two separate PRs. They are orthogonal changes.

aaronc commented:
> It may be good to keep the removal of bech32 and removing code duplication/test helper package as two separate PRs. They are orthogonal changes.
> 
> If someone needs to come back to this PR to view the changes having it only be the removal of bech32, like the title describes, will make the review quicker

I tend to agree with this perspective. I know it's annoying and I've had to regretfully split up many of my PRs for reasons like this. In the end though, I think this approach is the best for the SDK.

robert-zaremba commented:
> > It may be good to keep the removal of bech32 and removing code duplication/test helper package as two separate PRs. They are orthogonal changes.
> > If someone needs to come back to this PR to view the changes having it only be the removal of bech32, like the title describes, will make the review quicker
> 
> I tend to agree with this perspective. I know it's annoying and I've had to regretfully split up many of my PRs for reasons like this. In the end though, I think this approach is the best for the SDK.

Yes. I also agree, however when going through the code it was already in my head to refactor. I won't add more things. All this changes are related - all this edits would require doing extra error check. Instead, I moved it to a function.

robert-zaremba commented:
I'm breaking this PR into 2, for the moment I've extracted a code for test refactor and will submit another PR for it.

clevinson commented:
This probably also needs to include removing this from the keyring (cc @alexanderbez )

robert-zaremba commented on server/tm_cmds.go:
```diff
@@ -49,27 +47,19 @@ func ShowValidatorCmd() *cobra.Command {
 			}
-
-			fmt.Println(pubkey)
+			fmt.Println(string(out))
```
@aaronc , can you confirm if this makes sense?
+ we use the JSON protobuf marshaler for pubkey
+ shall we convert to string - as it is done here?

amaurymartiny replied:
Showing the pubkey Any as JSON, as you did in the latest commits, sgtm 👍

aaronc commented on codec/json.go:
```diff
@@ -28,3 +34,22 @@ func ProtoMarshalJSON(msg proto.Message, resolver jsonpb.AnyResolver) ([]byte, e
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "Expecting protobuf Message type, got %T", i)
+	}
+	return pm, nil
+}
```
Please revert these functions. This is not desirable. Just do a type assertion inline.

robert-zaremba replied:
moved to internal

blushi commented on codec/any.go:
```diff
@@ -14,7 +14,7 @@ import (
 // MarshalIfc is a convenience function for proto marshalling interfaces. It
 // packs the provided value in an Any and then marshals it to bytes.
-// NOTE: if you use a concret type, then you should use BinaryMarshaler.MarshalBinaryBare directly
+// NOTE: if you use a concert type, then you should use BinaryMarshaler.MarshalBinaryBare directly
```
should that be "concrete"?

robert-zaremba replied:
yes, not sure why I changed this :rofl:

alessio commented on types/bech32/legacybech32/pk_bench_test.go:
```diff
@@ -1,14 +1,12 @@
-	"github.com/stretchr/testify/require"
-
 	"github.com/cosmos/cosmos-sdk/crypto/keys/ed25519"
-	"github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
```
I think it may help to run `make format` here (I'd suggest to use the git hook available in `contrib`)

robert-zaremba replied:
Linter doesn't handle that (doesn't report this as an error). I fixed it. Thanks for catching.

robert-zaremba commented:
We have bunch of following linter issues :
```
SA1019: signing.SignMode_SIGN_MODE_LEGACY_AMINO_JSON is deprecated:
```
In files like: x/auth/tx/mode_handler.go:29, x/auth/ante/sigverify.go:185, client/tx/factory.go:45

I disabled `deprecation` check (SA1019) in all legacy files. Here, however, we are using it in a non - legacy file. 

My proposal is to add `//nolint:staticheck` in that lines. Comments?

robert-zaremba commented:
Summary:
+ removed bech32 from user facing code
+ updated the keyring display structure
+ changed user/client pubkey serialization
+ added new tests for `ed25519` as an example on how to properly serialize interfaces (both binary and JSON).
+ few API breaking changes: 
+ Created `NewJSONAnyMarshaler` - a helper method to use `codec.(Un)marsjalJSON` withing a client context:
```
am := codec.NewJSONAnyMarshaler(ctx.JSONMarshaler, ctx.InterfaceRegistry)
err := codec.UnmarshalIfcJSON(am, &pk, []byte(pubKey))
```

robert-zaremba commented:
See my comment here about followup tasks: https://github.com/cosmos/cosmos-sdk/issues/7447#issuecomment-735920200

robert-zaremba commented on crypto/keyring/info.go:
```diff
@@ -246,6 +232,7 @@ func (i multiInfo) UnpackInterfaces(unpacker codectypes.AnyUnpacker) error {
 // encoding info
 func marshalInfo(i Info) []byte {
+	// TODO: Why do we use Legacy Amino marshaling here?
 	return CryptoCdc.MustMarshalBinaryLengthPrefixed(i)
```
can anyone answer if we can use protobuf here?

amaurymartiny replied:
This is related to #7108. We marshal the Info as amino, and amino bytes are stored in the keyring backend.

robert-zaremba commented on server/tm_cmds.go:
```diff
@@ -133,6 +124,8 @@ against which this app has been compiled.
 
+// Deprecated: prints the content to the standard output using Legacy Amino
+// TODO: add issue to trace it?
 func printlnJSON(v interface{}) error {
```
shell we use protobuf here?

amaurymartiny replied:
Showing the pubkey Any as JSON, as you did in the latest commits, sgtm 👍

amaurymartiny commented on client/debug/main.go:
```diff
@@ -34,87 +33,32 @@ func Cmd() *cobra.Command {
-			cmd.Println("Bech32 Validator Consensus:", consenusPub)
-
+			cmd.Println("Address:", pk.Address())
+			cmd.Println("PubKey Hex:", hex.EncodeToString(pk.Bytes()))
```
I would propose to print here base64 instead of hex.

amaurymartiny commented on client/grpc/tmservice/service.go:
```diff
@@ -129,27 +114,31 @@ func (s queryServer) GetValidatorSetByHeight(ctx context.Context, req *GetValida
+		out[i] = &Validator{
+			Address:          v.Address,
+			ProposerPriority: v.ProposerPriority,
+			PubKey:           string(pkBz),
```
I just noticed that string here is not ideal, I created https://github.com/cosmos/cosmos-sdk/issues/8054

robert-zaremba replied:
Commented in the task. Let's handle it there.

amaurymartiny commented on client/keys/add_ledger_test.go:
```diff
@@ -122,6 +122,6 @@ func Test_runAddCmdLedger(t *testing.T) {
 	require.Equal(t,
-		"cosmospub1addwnpepqd87l8xhcnrrtzxnkql7k55ph8fr9jarf4hn6udwukfprlalu8lgw0urza0",
-		sdk.MustBech32ifyPubKey(sdk.Bech32PubKeyTypeAccPub, key1.GetPubKey()))
+		"PubKeySecp256k1{034FEF9CD7C4C63588D3B03FEB5281B9D232CBA34D6F3D71AEE59211FFBFE1FE87}",
```
I'm curious, any reason why you're using `.String()` here and `legacybech32.MustMarshalPubKey` above?

robert-zaremba replied:
This is a test. We just want to check if the content is good. Look at the 2 strings, which one is easier to reason about? For me the _new_ one because it shows the type name and easy to read hex sequence. Moreover we don't want to use `legacy`  code wherever possible.

amaurymartiny replied:
+1. So let's use `.String()` on line 78 too?

robert-zaremba replied:
I will see, I know that some `String` methods fail.

amaurymartiny commented on x/slashing/client/cli/query.go:
```diff
@@ -51,16 +52,16 @@ $ <appd> query slashing signing-info cosmosvalconspub1zcjduepqfhvwcmt7p06fvdgexx
 			params := &types.QuerySigningInfoRequest{ConsAddress: consAddr.String()}
-			res, err := queryClient.SigningInfo(context.Background(), params)
+			queryClient := types.NewQueryClient(clientCtx)
+			res, err := queryClient.SigningInfo(cmd.Context(), params)
```
the example in this cmd needs to be changed too.

robert-zaremba replied:
done:
{"@type":"/cosmos.crypto.ed25519.PubKey","key":"OauFcTKbN5Lx3fJL689cikXBqe+hcp6Y+x0rYUdR9Jk="}

amaurymartiny commented on x/slashing/client/cli/query.go:
```diff
@@ -51,16 +52,16 @@ $ <appd> query slashing signing-info cosmosvalconspub1zcjduepqfhvwcmt7p06fvdgexx
-			if err != nil {
+			var pk cryptotypes.PubKey
+			am := codec.NewJSONAnyMarshaler(clientCtx.JSONMarshaler, clientCtx.InterfaceRegistry)
+			if err := codec.UnmarshalIfcJSON(am, &pk, []byte(args[0])); err != nil {
```
thinking out loud, would it be easier to changes args[0] to valConsAddress as a bech32 address?

robert-zaremba replied:
Yes. Not sure if someone wants to query by a public key though. Should we check about it with someone else? Who?

amaurymartiny commented on server/tm_cmds.go:
```diff
@@ -91,7 +83,8 @@ func ShowAddressCmd() *cobra.Command {
 			if strings.ToLower(output) == "json" {
-				return printlnJSON(valConsAddr)
+				fmt.Printf("{\"address\": %q}\n", valConsAddr.String())
+				return nil
```
I think the --json flag should just print base64 bytes, if I'm not mistaken that's how it's done currently

robert-zaremba replied:
Yes. But `--json` flag should imply a JSON output, right? So here is a fix.

amaurymartiny replied:
I still don't understand why we need a JSON object `{"address":"..."}`. JSON doesn't necessarily mean JSON object, a string is also JSON.

But here I agree with you the `--json` flag is confusing. TBH i would remove this "if" block here, and always print bech32 string no matter the `-o` flag.

amaurymartiny commented on crypto/keyring/info.go:
```diff
@@ -176,35 +176,21 @@ func (i offlineInfo) GetPath() (*hd.BIP44Params, error) {
-	Threshold uint                 `json:"threshold"`
-	PubKeys   []multisigPubKeyInfo `json:"pubkeys"`
+	Name   string             `json:"name"`
+	PubKey cryptotypes.PubKey `json:"pubkey"`
```
I believe this struct is amino-marshalled and stored in the keyring backend, so this breaks backward-comptability of the keyring.

robert-zaremba replied:
It's used both in CLI and keyring storage. We are removing a fileds here, not changing it. It won't break unmarshalling.

amaurymartiny replied:
Are you sure about this?

If a user already has a keyring backend where the old multiInfo `bz` were stored, then, unmarshalling those `bz` into this new `multiInfo` will error.

robert-zaremba replied:
I will test it.

robert-zaremba replied:
Hmm, I thought I responded on this comment.
So... yes, you are right. `legacy.Cdc.UnmarshalBinaryLengthPrefixed` fails to decoded a subset of an encoded object. Thanks for catching. I added back the `multisigPubKeyInfo` and marked it deprecated. 
Also created an issue: https://github.com/cosmos/cosmos-sdk/issues/8323 to track the keyring migration.

robert-zaremba commented on CHANGELOG.md:
```diff
@@ -54,9 +54,17 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (x/staking) [\#7419](https://github.com/cosmos/cosmos-sdk/pull/7419) The `TmConsPubKey` method on ValidatorI has been removed and replaced instead by `ConsPubKey` (which returns a SDK `cryptotypes.PubKey`) and `TmConsPublicKey` (which returns a Tendermint proto PublicKey).
+* [\#7477](https://github.com/cosmos/cosmos-sdk/pull/7477) Changed Bech32 Public Key serialization in the client facing functionality (CLI, MsgServer, QueryServer):
+  * updated the keyring display structure (it uses protobuf JSON serialization) - the output is more verbose.
+  * Renamed `MarshalAny` and `UnmarshalAny` to `MarshalIfc` and `UnmarshalIfc` respectively. These functions must take interface as a parameter (not a concrete type nor `Any` object). Underneeth they use `Any` wrapping for correct protobuf serialization.
```
Updated in other PR, which I'm preparing for Stargate.

robert-zaremba commented on client/debug/main.go:
```diff
@@ -34,87 +33,32 @@ func Cmd() *cobra.Command {
+func getPubKeyFromString(ctx client.Context, pkstr string) (cryptotypes.PubKey, error) {
+	var pk cryptotypes.PubKey
+	am := codec.NewJSONAnyMarshaler(ctx.JSONMarshaler, ctx.InterfaceRegistry)
+	err := codec.UnmarshalIfcJSON(am, &pk, []byte(pkstr))
```
Good idea. I will do it in other PR.

amaurymartiny commented on client/debug/main.go:
```diff
@@ -10,13 +9,12 @@ import (
 	"github.com/cosmos/cosmos-sdk/version"
 )
 
+// Cmd creats a main CLI command
```
```suggestion
// Cmd creates a main CLI command
```

amaurymartiny commented on client/debug/main.go:
```diff
@@ -34,87 +33,32 @@ func Cmd() *cobra.Command {
 // getPubKeyFromString returns a Tendermint PubKey (PubKeyEd25519) by attempting
 // to decode the pubkey string from hex, base64, and finally bech32. If all
```
up

amaurymartiny commented on client/keys/add.go:
```diff
@@ -59,31 +59,31 @@ required through --multisig-threshold. The keys are sorted by address, unless
+	f.StringSlice(flagMultisig, nil, "Construct and store a multisig public key (implies --pubkey)")
+	f.Int(flagMultiSigThreshold, 1, "K out of N required signatures. For use in conjunction with --multisig")
+	f.Bool(flagNoSort, false, "Keys passed to --multisig are taken in the order they're supplied")
+	f.String(FlagPublicKey, "", "Parse a public key in bech32 format and save it to disk")
```
This needs to be updated

robert-zaremba replied:
Updated to: _Parse a public key in JSON format and saves key info to <name> file._

amaurymartiny commented on codec/codec.go:
```diff
@@ -69,3 +74,14 @@ type (
+
+// NewJSONAnyMarshaler creates a JSONAnyMarshaler using JSONMarshaler
+// and InterfaceRegistry
+func NewJSONAnyMarshaler(jm JSONMarshaler, ir types.InterfaceRegistry) IfcJSONMarshaler {
```
I believe all changes in this file are to be removed.

robert-zaremba replied:
Yes, these are old names before porting this to `master`. Git didn't flag it, but it's not used and we can remove it.

amaurymartiny commented on codec/json.go:
```diff
@@ -3,18 +3,23 @@ package codec
 	// We use the OrigName because camel casing fields just doesn't make sense.
 	// EmitDefaults is also often the more expected behavior for CLI users
-	jm := &jsonpb.Marshaler{OrigName: true, EmitDefaults: true, AnyResolver: resolver}
+	jm := defaultJM
+	if resolver != nil {
+		jm = &jsonpb.Marshaler{OrigName: true, EmitDefaults: true, AnyResolver: resolver}
+	}
```
what does this bring compared to the old (shorter) code?

robert-zaremba replied:
The default marshaler is cached. In many places we don't pass a resolver, so we can reuse a default one.

robert-zaremba replied:
Yes, you are right. `legacy.Cdc.UnmarshalBinaryLengthPrefixed` fails if we reduce the structure.
Thanks for catching. 

I've created an issue to trace it: https://github.com/cosmos/cosmos-sdk/issues/8323 (tentatively assigned to 0.42 milestone).

amaurymartiny commented on crypto/keyring/output.go:
```diff
@@ -1,106 +1,73 @@
-	Weight  uint   `json:"weight" yaml:"weight"`
+		Type:    keyType.String(),
+		Address: a.String(),
+		PubKey:  string(bz),
```
This will try to escape `"`s, e.g.

```json
{
  "name": "abc",
  "pub_key": "{\"@type\": \"...\"}"
}
```

I think for now it's fine, but ideally `KeyOutput` would be a proto message.

robert-zaremba replied:
Yes, but this is out of the scope. Main thing here was to change the PubKey type.

amaurymartiny commented on crypto/keys/ed25519/ed25519_test.go:
```diff
@@ -228,3 +231,80 @@ func TestMarshalAmino_BackwardsCompatibility(t *testing.T) {
+	require.True(pk3.Equals(pk))
+}
+
+func TestMarshalProto2(t *testing.T) {
```
It's very hard to understand the difference between the 2 tests, can you think of a better name?

robert-zaremba replied:
I will use TestMarshalProto and TestMarshalProtoInterface.
Feel free to share suggestions

amaurymartiny commented on crypto/keys/ed25519/ed25519_test.go:
```diff
@@ -228,3 +231,80 @@ func TestMarshalAmino_BackwardsCompatibility(t *testing.T) {
 	}
 }
+
+// TODO - maybe we should move the tests below to `codec_test` package, WDYT?
```
Sounds like a good idea 👍

robert-zaremba replied:
So, I'm not convinced any more (doesn't have a preference). This tests only deals with ed25519. On the other hand it's a good example for showing how to deal with interface encoding. 

Other idea is to link this test in a documentation. 

Anyone has a preference?

blushi replied:
ed25519 is just used as an example here so I think I'd prefer to put it in codec_test, although not a strong preference.

robert-zaremba replied:
OK, then let's do that. Moved to `/codec`

amaurymartiny reviewed (changes requested):
didn't finish the review yet, these are some first comments. My main concern is backwards-compatbility of keyring's multiInfo.

aaronc commented on CHANGELOG.md:
```diff
@@ -36,6 +36,21 @@ Ref: https://keepachangelog.com/en/1.0.0/
+
+* [\#7477](https://github.com/cosmos/cosmos-sdk/pull/7477) Changed Bech32 Public Key serialization in the client facing functionality (CLI, MsgServer, QueryServer):
+  * updated the keyring display structure (it uses protobuf JSON serialization) - the output is more verbose.
+  * Renamed `MarshalAny` and `UnmarshalAny` to `MarshalIfc` and `UnmarshalIfc` respectively. These functions must take interface as a parameter (not a concrete type nor `Any` object). Underneeth they use `Any` wrapping for correct protobuf serialization.
```
```suggestion
  * Renamed `MarshalAny` and `UnmarshalAny` to `MarshalInterface` and `UnmarshalInterface` respectively. These functions must take interface as a parameter (not a concrete type nor `Any` object). Underneeth they use `Any` wrapping for correct protobuf serialization.
```

aaronc commented on CHANGELOG.md:
```diff
@@ -36,6 +36,21 @@ Ref: https://keepachangelog.com/en/1.0.0/
+
+* (x/staking/types) [\#7447](https://github.com/cosmos/cosmos-sdk/issues/7447) Remove bech32 PubKey support:
+  * `ValidatorI` interface update. `GetConsPubKey` renamed to `TmConsPubKey` (consensus public key must be a tendermint key). `TmConsPubKey`, `GetConsAddr` methods return error.
+  * `Validator` update. Methods changed in `ValidatorI` (as described above) and `ToTmValidator` return error.
```
Can this be worded a bit more clearly?

robert-zaremba replied:
Updated.

aaronc commented on codec/codec.go:
```diff
@@ -48,6 +48,11 @@ type (
 		MustUnmarshalJSON(bz []byte, ptr proto.Message)
 	}
 
+	IfcJSONMarshaler interface {
```
```suggestion
	InterfaceJSONMarshaler interface {
```

aaronc reviewed (dismissed):
Let's please not use `Ifc` instead of `Interface` anywhere. It just feels super cryptic.

robert-zaremba commented:
I needed to add a new error type: `sdkerrors.ErrIO` and `sdkerros.AsOf` helper function: https://github.com/cosmos/cosmos-sdk/pull/7477/commits/8eba4b021b93aceb0a501187af9e7cfd1b013524

If needed I can cherry pick it into a new PR.

blushi commented on codec/proto_codec.go:
```diff
@@ -155,6 +169,8 @@ func (pc *ProtoCodec) UnmarshalJSON(bz []byte, ptr proto.Message) error {
 
 // MustUnmarshalJSON implements JSONMarshaler.MustUnmarshalJSON method,
 // it executes UnmarshalJSON except it panics upon failure.
+// NOTE: this function must be used with a concert type which
```
```suggestion
// NOTE: this function must be used with a concrete type which
```

blushi commented on crypto/keys/ed25519/ed25519_test.go:
```diff
@@ -228,3 +230,81 @@ func TestMarshalAmino_BackwardsCompatibility(t *testing.T) {
+	var pkAny2 codectypes.Any
+	err = ccfg.Marshaler.UnmarshalJSON(bz, &pkAny2)
+	require.NoError(err)
+	// we before getting a cached value we need to unpack it.
```
```suggestion
	// Before getting a cached value we need to unpack it.
```

blushi commented on crypto/keys/ed25519/ed25519_test.go:
```diff
@@ -228,3 +230,81 @@ func TestMarshalAmino_BackwardsCompatibility(t *testing.T) {
+	err = ccfg.Marshaler.UnmarshalJSON(bz, &pkAny2)
+	require.NoError(err)
+	// we before getting a cached value we need to unpack it.
+	// Normally this happens in in types which implement UnpackInterfaces
```
```suggestion
	// Normally this happens in types which implement UnpackInterfaces.
```

blushi commented on crypto/keys/ed25519/ed25519_test.go:
```diff
@@ -228,3 +230,81 @@ func TestMarshalAmino_BackwardsCompatibility(t *testing.T) {
+	require.True(pk3.Equals(pk))
+
+	// ** Check unmarshal using JSONMarshaler **
+	// Unpacking won't work straightforward s Any type
```
```suggestion
	// Unpacking won't work straightforward with Any type
```

blushi commented on crypto/keys/ed25519/ed25519_test.go:
```diff
@@ -228,3 +230,81 @@ func TestMarshalAmino_BackwardsCompatibility(t *testing.T) {
+
+	// ** Check unmarshal using JSONMarshaler **
+	// Unpacking won't work straightforward s Any type
+	// Any can't implement UnpackInterfacesMessage insterface. So Any is not
```
```suggestion
	// Any can't implement UnpackInterfacesMessage interface. So Any is not
```

blushi commented on crypto/keys/multisig/multisig.go:
```diff
@@ -16,18 +16,18 @@ var _ types.UnpackInterfacesMessage = &LegacyAminoPubKey{}
 // Panics if len(pubKeys) < k or 0 >= k.
-func NewLegacyAminoPubKey(k int, pubKeys []cryptotypes.PubKey) *LegacyAminoPubKey {
-	if k <= 0 {
+func NewLegacyAminoPubKey(treshold int, pubKeys []cryptotypes.PubKey) *LegacyAminoPubKey {
```
```suggestion
func NewLegacyAminoPubKey(threshold int, pubKeys []cryptotypes.PubKey) *LegacyAminoPubKey {
```

blushi commented on server/tm_cmds.go:
```diff
@@ -29,11 +29,9 @@ func ShowNodeIDCmd() *cobra.Command {
 
 			nodeKey, err := p2p.LoadOrGenNodeKey(cfg.NodeKeyFile())
-			if err != nil {
-				return err
```
why not returning the error anymore?

blushi commented on x/slashing/simulation/decoder_test.go:
```diff
@@ -47,16 +48,15 @@ func TestDecodeStore(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
-			switch i {
-			case len(tests) - 1:
+			if i == len(tests)-1 {
```
can't we just add some `expectPanic` field to the tests struct instead?

blushi commented on x/staking/client/cli/cli_test.go:
```diff
@@ -1,4 +1,4 @@
-// +build norace
+// build norace
```
why removing this?

blushi reviewed (commented):
Looks good to me overall, just a few minor comments/typos.

amaurymartiny commented on crypto/keyring/info.go:
```diff
@@ -191,21 +195,14 @@ type multiInfo struct {
-	}
+		Name:   name,
+		PubKey: pub,
+	}, nil
```
shouldn't we keep the 2 deleted fields, following https://github.com/cosmos/cosmos-sdk/pull/7477#discussion_r556534909?

amaurymartiny commented on server/tm_cmds.go:
```diff
@@ -49,31 +47,24 @@ func ShowValidatorCmd() *cobra.Command {
 		},
 	}
 
-	cmd.Flags().StringP(cli.OutputFlag, "o", "text", "Output format (text|json)")
```
This is actually a breaking change. Either add an entry to the changelog, or we could not make it a breaking change by:
- `--json`: ouput proto JSON
- `--text`: output sdkPk.String()

amaurymartiny commented on client/keys/show.go:
```diff
@@ -144,19 +146,18 @@ func runShowCmd(cmd *cobra.Command, args []string) (err error) {
 func fetchKey(kb keyring.Keyring, keyref string) (keyring.Info, error) {
+	// firstly check if the keyref is a key name of a key registered in a keyring.
 	info, err := kb.Key(keyref)
+	if err == nil || !sdkerr.AsOf(err, sdkerr.ErrIO, sdkerr.ErrKeyNotFound) {
+		return info, err
+	}
```
Can you explain why we need `|| !sdkerr.AsOf(err, sdkerr.ErrIO, sdkerr.ErrKeyNotFound)` ? Maybe a comment would be good here.

amaurymartiny commented on types/errors/errors.go:
```diff
@@ -135,6 +135,10 @@ var (
 
+	// ErrIO should be used to wrap internal errors caused by external operation.
+	// Examples: not DB domain error, file writing etc...
+	ErrIO = Register(RootCodespace, 38, "Internal IO error")
```
it seems to me no-one is ever throwing this error. Would it make sense to introduce it when we need it?

robert-zaremba replied:
I'm using it for differentiating an internal keyring error, from other errors.

amaurymartiny reviewed (commented):
I'm still not sure we should remove the 2 fields in multi info. The other comments are small nits.

odeke-em commented on client/keys/add.go:
```diff
@@ -186,16 +187,13 @@ func RunAddCmd(cmd *cobra.Command, args []string, kb keyring.Keyring, inBuf *buf
-		}
-
-		return nil
+		_, err := kb.SavePubKey(name, pk, algo.Name())
```
Does this code compile?

robert-zaremba replied:
yes, why not?

odeke-em commented on types/errors/errors.go:
```diff
@@ -358,6 +362,17 @@ func WithType(err error, obj interface{}) error {
 
+// IsOf checks if a received error is caused by one of the target errors.
+// It extends the errors.Is functionality to a list of errors.
+func IsOf(received error, targets ...error) bool {
```
We could further improve this by declaring its intent as: AnyOf or OfAny. We could also improve the comment and make it more precise by saying that the function returns true if any of the targets satisfy errors.Is for the subject error.

robert-zaremba replied:
But those are two different functions - we would need to create `AnyOf`. This is what you have in mind?
Ref: https://pkg.go.dev/errors#As

odeke-em commented on types/errors/errors_test.go:
```diff
@@ -149,6 +149,27 @@ func (s *errorsTestSuite) TestErrorIs() {
+
+	require.True(IsOf(errW, ErrLogic))
+	require.True(IsOf(errW, err, ErrLogic))
+	require.True(IsOf(errW, nil, errW), "error should much itsel")
```
“itself”

odeke-em reviewed (commented):
Thank you @robert-zaremba, just a drive-by review that was intended for security.

clevinson reviewed (commented):
Few questions about stuff in the keys command that affects multisig UX.

In particular, `simd keys show --pubkey` does not work, and `simd keys show` does not provide a proper JSON econded Any value. A proper Any value (with `@type` field) is necessary for passing this along to the `simd keys add --pubkey` for later use in the `simd keys add --multisig {NAMES} --multisig-threshold {N}` command.

I tried looking into the code to see why the `show` command does not actually show the full JSON Any value (with `@type` field). Are you able to add that in to the printing? Or was this excluded on purpose?

robert-zaremba commented on client/keys/show.go:
```diff
@@ -111,10 +115,8 @@ func runShowCmd(cmd *cobra.Command, args []string) (err error) {
-	case isShowPubKey:
-		printPubKey(cmd.OutOrStdout(), info, bechKeyOut)
+	case isShowAddr, isShowPubKey:
+		printInfo(cmd.OutOrStdout(), info, bechKeyOut)
```
Correct - there was a bug.

robert-zaremba commented on crypto/keyring/output.go:
```diff
@@ -1,106 +1,73 @@
 // Bech32KeyOutput create a KeyOutput in with "acc" Bech32 prefixes. If the
 // public key is a multisig public key, then the threshold and constituent
 // public keys will be added.
 func Bech32KeyOutput(keyInfo Info) (KeyOutput, error) {
```
> `simd keys show` does not provide a proper JSON econded Any value.

We never presented an Any value here. Before (in master):
```
% ./build/simd keys show node0
- name: node0
  type: local
  address: cosmos1ye4mulc85xqmn5s7dg9rhk0n3hv8z6m5e5ll6n
  pubkey: cosmospub1addwnpepq2qm7zqv86dsrjdhj3kesv0hzr3gf2ndxttlhz9xcjc93a9xrasv6hn902c
  mnemonic: ""
  threshold: 0
  pubkeys: []

% ./build/simd keys show node0 -p
cosmospub1addwnpepq2qm7zqv86dsrjdhj3kesv0hzr3gf2ndxttlhz9xcjc93a9xrasv6hn902c
```
After (in this branch):
```
% ./build/simd keys show node0   
- name: node0
  type: local
  address: cosmos1ye4mulc85xqmn5s7dg9rhk0n3hv8z6m5e5ll6n
  pubkey: '{"key":"AoG/CAw+mwHJt5RtmDH3EOKEqm0y1/uIpsSwWPSmH2DN"}'
  mnemonic: ""

% ./build/simd keys show node0 -p
{"key":"AoG/CAw+mwHJt5RtmDH3EOKEqm0y1/uIpsSwWPSmH2DN"}
```

robert-zaremba replied:
Yes, good idea. I will rename.

robert-zaremba replied:
Ah, the function returns a `KeyOutput`, which contains an `Address` which is formatted using bech32.
That being said, I think we don't need bech32 prefix in this function names.

robert-zaremba replied:
Renamed: `Bech32...` -> `Fmt...`

robert-zaremba commented:
> `simd keys show` does not provide a proper JSON econded Any value.

We never presented an Any value here. Before (in master):
```
% ./build/simd keys show node0
- name: node0
  type: local
  address: cosmos1ye4mulc85xqmn5s7dg9rhk0n3hv8z6m5e5ll6n
  pubkey: cosmospub1addwnpepq2qm7zqv86dsrjdhj3kesv0hzr3gf2ndxttlhz9xcjc93a9xrasv6hn902c
  mnemonic: ""
  threshold: 0
  pubkeys: []

% ./build/simd keys show node0 -p
cosmospub1addwnpepq2qm7zqv86dsrjdhj3kesv0hzr3gf2ndxttlhz9xcjc93a9xrasv6hn902c
```
After (in this branch):
```
% ./build/simd keys show node0   
- name: node0
  type: local
  address: cosmos1ye4mulc85xqmn5s7dg9rhk0n3hv8z6m5e5ll6n
  pubkey: '{"key":"AoG/CAw+mwHJt5RtmDH3EOKEqm0y1/uIpsSwWPSmH2DN"}'
  mnemonic: ""

% ./build/simd keys show node0 -p
{"key":"AoG/CAw+mwHJt5RtmDH3EOKEqm0y1/uIpsSwWPSmH2DN"}
```

@clevinson

alessio commented on crypto/keyring/output.go:
```diff
@@ -1,106 +1,73 @@
-	}
-
-	return ko, nil
+func MkKeyOutput(keyInfo Info) (KeyOutput, error) {
```
```suggestion
func MakeKeyOutput(keyInfo Info) (KeyOutput, error) {
```

may be more readable

robert-zaremba replied:
I like `Mk` as well - it's meaning is pretty obvious and makes the code slightly shorter.

amaurym replied:
What about 
```suggestion
func MkAccKeyOutput(keyInfo Info) (KeyOutput, error) {
```

? see https://github.com/cosmos/cosmos-sdk/pull/7477#discussion_r592298247

robert-zaremba replied:
Yeap. Good suggestion.

alessio commented on crypto/keyring/output.go:
```diff
@@ -1,106 +1,73 @@
-// call to Bech32KeyOutput fails.
-func Bech32KeysOutput(infos []Info) ([]KeyOutput, error) {
+// call to MkKeyOutput fails.
+func MkKeysOutput(infos []Info) ([]KeyOutput, error) {
```
```suggestion
func MakeKeysOutput(infos []Info) ([]KeyOutput, error) {
```

alessio replied:
for readability

robert-zaremba replied:
I was trying to find even shorter name (I think `mk` is pretty readable and in unix spirit).

alessio reviewed (commented):
I'd suggest to move `MkKeysOutput` and `MkKeyOutput` to the package `client/keys` and un-export them. They're defined in `crypto/keyring` but they are actually used in `client/keys`.

robert-zaremba commented:
> I'd suggest to move `MkKeysOutput` and `MkKeyOutput` to the package `client/keys` and un-export them. They're defined in `crypto/keyring` but they are actually used in `client/keys`.

@alessio - yes - I also wanted to do it, but it was creating some conflicts in the past, so I created a task for it  #8046

amaurym commented on crypto/keyring/output.go:
```diff
@@ -1,106 +1,73 @@
+	Name     string `json:"name" yaml:"name"`
+	Type     string `json:"type" yaml:"type"`
+	Address  string `json:"address" yaml:"address"`
+	PubKey   string `json:"pubkey" yaml:"pubkey"`
```
```suggestion
	PubKey   *Any `json:"pubkey" yaml:"pubkey"`
```

related to https://github.com/cosmos/cosmos-sdk/pull/7477#pullrequestreview-608323873, I would also prefer `Any` here, so that 1. the key type is shown, and 2. users can copy-paste

amaurym replied:
> We could update the format - but that will require passing a codec down to the keyring functions.

We don't need, we just need to do `pkAny := codec.NewAnyWithValue(pk)` in the `NewKeyOutput()` function

robert-zaremba replied:
`KeyOutput` is meant to be used for direct printing (struct fields printing) or printing use json/yaml serialization. Calling `MarshalJSON` for type *types.Any is not supported and fails.

amaurym replied:
Ah, good point, let's keep `PubKey string` then

robert-zaremba replied:
> let's keep PubKey string then, but the string is the JSON repr of the Any.

Yes. 
The `KeyOutput` can be used with any output serialization, including yaml (eg `printTextInfos` function)

amaurym reviewed (commented):
lgtm overall, thanks @robert-zaremba. I'm echoing @clevinson's concern: I would also see the KeyOutput use Any for the pubkey field.

blushi commented on codec/any_test.go:
```diff
@@ -53,3 +57,80 @@ func TestMarshalAny(t *testing.T) {
+	var pkAny2 codectypes.Any
+	err = ccfg.Marshaler.UnmarshalJSON(bz, &pkAny2)
+	require.NoError(err)
+	// we before getting a cached value we need to unpack it.
+	// Normally this happens in in types which implement UnpackInterfaces
```
```suggestion
	// Before getting a cached value we need to unpack it.
	// Normally this happens in types which implement UnpackInterfaces.
```

blushi commented on crypto/ledger/ledger_test.go:
```diff
@@ -102,26 +102,15 @@ func TestPublicKeySafe(t *testing.T) {
 }
 
 func TestPublicKeyHDPath(t *testing.T) {
+	// TODO: convert this tests to not use bech32
```
should we add an issue for that or reference an existing issue here?

robert-zaremba replied:
In fact I'm solving it here.

robert-zaremba replied:
I'm removing the todo, but we should run the test with ledger support. What's the best way to do it?

blushi replied:
what do you mean exactly by "with ledger support"?

robert-zaremba replied:
if we can have a test with a real ledger

robert-zaremba replied:
I will update this test in a separate PR.

blushi replied:
ok do we have an issue to track that?

blushi reviewed (commented):
Looks good to me overall but I agree with @amaurym and @clevinson that we should show the pubkey values in the CLI.

robert-zaremba commented on x/auth/testutil/suite.go:
```diff
@@ -226,7 +226,7 @@ func (s *TxConfigTestSuite) TestTxEncodeDecode() {
 		PubKey: pubkey,
 		Data: &signingtypes.SingleSignatureData{
-			SignMode:  signingtypes.SignMode_SIGN_MODE_LEGACY_AMINO_JSON,
+			SignMode:  signingtypes.SignMode_SIGN_MODE_LEGACY_AMINO_JSON, //nolint:staticcheck
```
SignMode_SIGN_MODE_LEGACY_AMINO_JSON is marked as deprecated. Adding new code which uses a deprecated features makes linter to complain.

blushi replied:
it seems like there are still places where it needs to be added, lint is failing

robert-zaremba replied:
linter is failiing because of the SignMode_SIGN_MODE_LEGACY_AMINO_JSO usage in a modified files.

amaurym commented on CHANGELOG.md:
```diff
@@ -186,6 +187,24 @@ he Cosmos Hub) should not use this release or any release in the v0.41.x series.
+
+### API Breaking
+
+* (x/staking/types) [\#7447](https://github.com/cosmos/cosmos-sdk/issues/7447) Remove bech32 PubKey support:
```
up

amaurym reviewed (approved):
tACK, thanks @robert-zaremba, this is great!

I tested `simd keys list`, and the pubkey output is correct proto JSON.

marbar3778 commented on client/keys/codec.go:
```diff
@@ -5,6 +5,8 @@ import (
 	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
 )
 
+// TODO: remove this file #8047
```
nit: post whole issue url to easily access

robert-zaremba replied:
ok

marbar3778 reviewed (approved):
Generally looks good. The PR has a bit of scope creep but should be fine

blushi reviewed (approved):
Tested ACK!

robert-zaremba commented:
Any idea why the following test is failing so often?
```
 --- FAIL: TestPaginatedVotesQuery (0.00s)
    querier_test.go:336: 
        	Error Trace:	querier_test.go:336
        	Error:      	Not equal: 
        	            	expected: 19
        	            	actual  : 20
        	Test:       	TestPaginatedVotesQuery
FAIL
```

alessio commented:
> Any idea why the following test is failing so often?
> 
> ```
>  --- FAIL: TestPaginatedVotesQuery (0.00s)
>     querier_test.go:336: 
>         	Error Trace:	querier_test.go:336
>         	Error:      	Not equal: 
>         	            	expected: 19
>         	            	actual  : 20
>         	Test:       	TestPaginatedVotesQuery
> FAIL
> ```

Just a flaky test. But it's annoying and no one can really explain why this happens in the first. We definitely need to fix it.

robert-zaremba commented:
yeah, finally merged after 20 attempts to get the CI go through random errors.
